### PR TITLE
Adding Backward Support for NestedTensors and FlashAttention

### PR DESCRIPTION
**Pull request by drisspg**
# Summary

### <samp>🤖 Generated by Copilot at 318764f</samp>

This pull request implements the CUDA backend of the SDPA kernel for nested tensors, which enables efficient transformer models with variable-length sequences. It adds a new dispatch key, a backward function, a unit test, and some helper functions for the kernel. It modifies `test/test_transformers.py`, `aten/src/ATen/native/native_functions.yaml`, `aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp`, and `aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.h`.

### <samp>🤖 Generated by Copilot at ed4a773</samp>

> _Fused kernels of doom, unleash the flash attention_
> _Nested tensors on fire, reshape and pad with caution_
> _Backward pass of power, dispatch the CUDA key_
> _Test the gradients of hell, warn the user if they disagree_

cc @cpuhrsch @jbschlosser @bhosmer @soulitzer @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  grad_v = wrap_buffer(grad_v.view(-1), value._nested_tensor_size());
+
+  return std::make_tuple(grad_q, grad_k, grad_v);
+}
+
+} // namespace native
+} // namespace at
+
+
+//     const Tensor& query,
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp:75]: Deleeeeeeeeeetes

**Review by alexsio27444 — CHANGES_REQUESTED** [review]
Change

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    cumulative_sequence_length_k,
+    max_seqlen_batch_q,
+    max_seqlen_batch_k,
+    dropout_p,
+    is_causal,
+    philox_seed,
+    philox_offset,
+    scale);
+
+  grad_q = wrap_buffer(grad_q.view(-1), query.transpose(1,2)._nested_tensor_size()).transpose(1,2);
```
[the remark this answers is not in the document]
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp:64]: why?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * @return A tuple containing all the necessary data for running the fused
+ * kernels
+ */
+std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor, int64_t, int64_t, Tensor>
+sdpa_nested_preprocessing(
+    const Tensor& query,
+    const Tensor& key,
+    const Tensor& value);
+
+std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor>
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.h:21]: add docstring for this

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+   * use with the flash-attention and efficient_attention kernels without
+   * needing to call contiguous on the nested tensor input.
+   * It checks that the storage offsets' adjacent_differences are a constant
+   * mutiple of the previous tensor in the nested tensor and that the strides
+   * are monitonically decreasing. This check is done after calling transpose on
+   * the nested tensor.
+   *
+   * @return A boolean indicating of contiguous needs to be called for input
+   */
+  bool is_safe_to_get_storage_as_tensor(const NestedTensorImpl* tensor) {
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:87]: unchanged just moved

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+   * (1) get the storage of the contiguous nested tensor
+   * (2) view as shape {output_batch_size, {*}_t.size(1), output_num_heads,
+   * head_dim_{*}}, and stride {0, nnz_{*}_stride, head_{*}_stride,
+   * head_dim_stride} where head_{*}_stride is 0 if
+   * {*}_num_heads_needs_broadcast (3) collapse the first two dims by reshaping
+   * to {Nnz_{*}, output_num_heads, head_dim_{*}} if {*}_t.size(1) (i.e. the
+   * seq_len is 1), the reshape should be a view and should not incur a copy
+   *  dense tensor without getting the storage
+   */
+  at::Tensor view_as_dense(
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:155]: I abstracted this out since we need it for both the forward preprocessing(broadcasted or not) and backwards preprocessing, I also think it makes it much cleaner

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+      const auto* query_impl = get_nested_tensor_impl(q_t);
+      if (!q_t.is_contiguous() &&
+          !is_safe_to_get_storage_as_tensor(query_impl)) {
+        q_t = q_t.contiguous();
+      }
+      // If we are broadcasting then Nnz_q will be the output_batch_size since
+      // seq_len is 1
+      const int64_t effective_batch_size_q =
+          q_batch_size_needs_broadcast ? output_batch_size : Nnz_q;
+      query_buffer_reshaped = view_as_dense(
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:297]: the view_as_dense are the only changes I made to this file

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+    q_t = q_t.contiguous();
+  }
+  if (!k_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(key_impl)) {
+    k_t = k_t.contiguous();
+  }
+  if (!v_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(value_impl)) {
+    v_t = v_t.contiguous();
+  }
+
+  query_buffer_reshaped = view_as_dense(q_t, Nnz_q, num_heads, head_dim_qk);
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:438]: only changes

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+    const at::Tensor& grad_out_,
+    const at::Tensor& query,
+    const at::Tensor& key,
+    const at::Tensor& value,
+    const at::Tensor& out,
+    const Tensor& cumulative_sequence_length_q,
+    const Tensor& cumulative_sequence_length_kv,
+    const int64_t max_seqlen_batch_q,
+    const int64_t max_seqlen_batch_kv) {
+//   const int64_t q_batch_size = query.size(0);
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:469]: Need to handle the broadcast path

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+
+ grad_out_buffer_reshaped =
+      view_as_dense(grad_out_t, Nnz_q, num_heads, head_dim_v);
+  output_buffer_reshaped = view_as_dense(out_t, Nnz_q, num_heads, head_dim_v);
+
+  auto output_shape = get_nested_sizes(q_t).clone();
+  if (head_dim_v != head_dim_qk) {
+    output_shape.select(1, -1).fill_(head_dim_v);
+  }
+  // MONDAY DRISS LISTEN TO ME:
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:541]: Monday me has listened and can remove this

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+ */
+int64_t get_nnz(Tensor nestedtensor) {
+  auto* nt_impl = get_nested_tensor_impl(nestedtensor);
+  const auto& sizes = nt_impl->get_nested_sizes();
+  auto size_tensor_stride = sizes.stride(0);
+  const int64_t batch_size = nestedtensor.size(0);
+  auto* sizes_ptr = sizes.data_ptr<int64_t>();
+  int64_t cumulative_sequence_length = 0;
+  for (const auto i : c10::irange(batch_size)) {
+    // Calculate the cumulative sum of the sequence lengths
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:26]: I could also use storage_offsets with some math to divide out by the n_heads and head_dim... not sure though if it is worth it/ makes it a little more abstract

**Comment by bhack** [comment]
Why this is stale?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 11]
```diff
@@ -449,6 +472,100 @@ sdpa_nested_preprocessing(
       output_shape);
 }
 
+std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor>
+sdpa_nested_preprocessing_backward(
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:476]: dumb question: why do we need separate preprocessing for forward and backward? I'd kinda expect the same required form for both

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 12]
```diff
@@ -449,6 +472,100 @@ sdpa_nested_preprocessing(
...
+
+  const int64_t num_heads = query.size(1);
+  const int64_t head_dim_qk = query.size(3);
+  const int64_t head_dim_v = value.size(3);
+
+  Tensor q_t = query.transpose(1, 2);
+  Tensor k_t = key.transpose(1, 2);
+  Tensor v_t = value.transpose(1, 2);
+  Tensor grad_out_t = grad_out_.transpose(1, 2);
+  Tensor out_t = out.transpose(1, 2);
```
Comment (jbschlosser) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:508]: only 4D shapes supported, right?

**Review by jbschlosser — APPROVED** [review]
Looking good! Some minor comments but nothing showstopping

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 13]
```diff
@@ -50,6 +50,73 @@ std::array<SDPBackend, num_backends> priority_order(sdp_params params) {
   return default_order;
 }
 
+bool check_batch_size_and_num_heads(sdp_params params, bool debug) {
```
[the remark this answers is not in the document]
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:53]: ohhh good catch this was likely nasty rebase bug

File: test/test_transformers.py [hunk 14]
```diff
@@ -137,6 +137,14 @@ def _size(i):
         size = (batch, seq_len, num_heads, head_dim) if not packed else (batch, seq_len, 3 * num_heads * head_dim)
         return torch.randn(size, device=device, dtype=dtype, requires_grad=requires_grad)
 
+def calculate_nt_tolerances(nt_ref_hp, nt_ref_lp, default_dtype, fudge_factor=1):
+    # TODO use NT ops when we have implemented Max for NestedTensor instead of unrolling
+    ref_atol = default_atol[default_dtype]
+    ref_rtol = default_rtol[default_dtype]
+    for tensor_component_ref, tensor_component_ref_lp in zip(nt_ref_hp.unbind(), nt_ref_lp):
```
Comment (drisspg) [inline test/test_transformers.py:144]: nt_ref_lp.unbind()*

File: test/test_transformers.py [hunk 15]
```diff
@@ -2836,6 +2844,116 @@ def test_fused_kernels_nested_broadcasting_query_dense(self, device):
 
         self.assertEqual(actual.contiguous(), math_ref.contiguous(), atol=1e-3, rtol=1e-2)
 
+    @onlyCUDA
```
[the remark this answers is not in the document]
Comment (drisspg) [reply test/test_transformers.py:2847]: I can resolve this since we talked about this online right?

[the remark this answers is not in the document]
Comment (jbschlosser) [reply test/test_transformers.py:2847]: yep! for posterity: the fused kernels don't work with float64, so gradcheck is out

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp:64]: This description is accurate. At least at this level grad_q, grad_k, grad_v are created within at_flash backward so I know the stride information here. and prior to calling at_flash_forward, nts must be layed out , N, seq_len, num_heads, D. SO I think it should be okay

File: test/test_transformers.py (hunk 14 above)
Comment (jbschlosser) [reply test/test_transformers.py:144]: `unbind()` happens implicitly when you foreach over it, no?

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+  grad_k = wrap_buffer(grad_k.view(-1), key.transpose(1,2)._nested_tensor_size()).transpose(1,2);
```
[the remark this answers is not in the document]
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctionsBackward.cpp]: Since for now we are asserting the input is 4d for flash_froward I think we should leave as is for now

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp (hunk 11 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:476]: we don't really need separate preprocessing it is really just to also apply the same processing on out, and grad_out

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp (hunk 12 above)
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:508]: yup only 4d shapes

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 17]
```diff
@@ -449,6 +472,100 @@ sdpa_nested_preprocessing(
...
+  if (!grad_out_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(grad_out_impl)) {
+    grad_out_t = grad_out_t.contiguous();
+  }
+  if (!out_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(out_impl)) {
+    out_t = out_t.contiguous();
+  }
+
+  query_buffer_reshaped = view_as_dense(q_t, Nnz_q, num_heads, head_dim_qk);
+  key_buffer_reshaped = view_as_dense(k_t, Nnz_kv, num_heads, head_dim_qk);
+  value_buffer_reshaped = view_as_dense(v_t, Nnz_kv, num_heads, head_dim_v);
```
[the remark this answers is not in the document]
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:546]: bingo

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp [hunk 18]
```diff
@@ -449,6 +472,100 @@ sdpa_nested_preprocessing(
...
+
+ grad_out_buffer_reshaped =
+      view_as_dense(grad_out_t, Nnz_q, num_heads, head_dim_v);
+  output_buffer_reshaped = view_as_dense(out_t, Nnz_q, num_heads, head_dim_v);
+
+  auto output_shape = get_nested_sizes(q_t).clone();
+  if (head_dim_v != head_dim_qk) {
+    output_shape.select(1, -1).fill_(head_dim_v);
+  }
```
[the remark this answers is not in the document]
Comment (drisspg) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp:555]: The last dim for k and v need not be the same for SDPA, and efficient attention supports this. Since we are only adding train support for flash right now it is not used but if we applied the same for mem eff it would be useful

File: test/test_transformers.py [hunk 19]
```diff
@@ -2836,6 +2844,116 @@ def test_fused_kernels_nested_broadcasting_query_dense(self, device):
...
+
+        query_ref = query.clone().detach().to(torch.float32).requires_grad_(True)
+        key_ref = key.clone().detach().to(torch.float32).requires_grad_(True)
+        value_ref = value.clone().detach().to(torch.float32).requires_grad_(True)
+
+        is_dropout = dropout_p > 0.0
+
+        if not is_dropout:
+            with sdp_kernel(enable_math=False, enable_flash=True, enable_mem_efficient=False):
+                out = F.scaled_dot_product_attention(query, key, value, dropout_p=dropout_p, is_causal=False, scale=scale)
```
[the remark this answers is not in the document]
Comment (drisspg) [reply test/test_transformers.py]: added, but we currently dont support this for NTs

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting you change, but one of the tests test_fused_kernels_nested_broadcasting_requires_grad_failure_cuda is failing on Windows CUDA https://hud.pytorch.org/pytorch/pytorch/commit/f7ba3e85e26927ab5296737186ec32ef06c69a69' -c nosignal

As the PR is still in diff train and not yet landed in fbcode, I think it would be easier to revert and reland this change.  I'll add `ciflow/periodic` to run the Windows CUDA jobs.

**Comment by drisspg** [comment]
Sounds good thank you @huydhn
